### PR TITLE
Implement provider for chrome debugging protocol (closes #1417)

### LINKED ISSUE #1417: Implement Chrome Debugging Protocol provider
### Are you requesting a feature or reporting a bug?
Feature

### What is the current behavior?
Currently there is no support for Chrome Debugging Protocol in TestCafe.

### What is the expected behavior?
Implement a provider that will use CDP  and will support:

 - [ ] Headless mode: screenshots, closing windows #1411

 - [ ] Device emulation #1274

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @inikulin @AlexanderMoskovkin

### CODE REVIEW

**Comment by AndreyBelym** [comment]
# Feature summary
## Usage
`testcafe 'cdp:config' test.js`
E. g.
`testcafe 'cdp:local;headless;device.name=iPhone 4' test.js`
`testcafe 'cdp:local;headless;device={name=iPhone 4;orientation=horizontal}' test.js`
## Config
    local:      Indicates that a local browser instance should be started by provider. Also triggered automatically by specifying path or arguments.
    args:      Optional user-specified arguments 
    incognito:  Indicates that a local browser instance should be started in incognito mode,
    headless:  Indicates that a local browser instance should be started in incognito mode,
    host:      Specifies a host to connect via CDB. May be specified with port, e.g. localhost:9222 ,
    port:      Specifies a port to connect via CDB,
    path:      Path to the chrome browser,

    userDataDir:    Specifies temporary directory for storing user profile to allow multiple running Chrome instances,
    noTempUserData: Don't create temporary user profile directory,

    emulation: Should enable DevTools Device Emulation,

    device: {
        screen: {
            width:   Device screen width,
            height:  Device screen height,
            density: Device screen physical/logical pixel ratio,
        },

        mobile:      Specifies that device is mobile,
        orientation: 'vertical' or 'horizontal',
        name:        Use named device preset from DevTools configuration,

        userAgent:      Specifies device useragent string,

        touch:          Device has touch features
    }

**Comment by inikulin** [comment]
Why we need separate `cdp` alias, can we use `chrome` instead, e.g. `chrome:headless`?

**Comment by AndreyBelym** [comment]
FPR

**Comment by AndreyBelym** [comment]
> > remote: Indicates that a local browser instance should not be started by provider.

> Why we need this. Isn't specifying host enables remote debugging

If you've started browser with CDB manually (on the port 9222), you can just write `remote` instead of `host=localhost:9222`. Also, you can use `remote` to connect Chrome on Android if you have a device connected by USB.

> > noCdb: Start local browser instance without Chrome Debugging Protocol support (no headless, no emulation)

> Why we need this in the first place. isn't it the same as testcafe chrome test.js?

I would like that CDB will be enabled by default, because it gives better screenshots: even if Chrome is in background, and without browser UI, on all platfroms (currently we don't include Chrome UI in screenshots only on Windows).

> > args: Optional user-specified arguments

> AFAIK we already have the way to specify args for providers, so why we need this?

You can specify arguments in the old way if you don't use any other options. Without it it will be harder to extract provider-specific options from user-supplied args. So if you don't plan to use emulation/headless/etc. you can still use
```
testcafe 'chrome --my-arg --my-arg2'
```
But if you are using emulation, you will use `args` option (live example from V. Gein):
```
testcafe "chrome:args=--disable-infobars --disable-session-crashed-bubble --kiosk;device.name=iPhone 
 4"
```
>  > host: Specifies a host to connect via CDB. May be specified with port, e.g. localhost:9222 , port: Specifies a port to connect via CDB

> Having a host is just enough, moreover it makes sense to name it `remote, e.g.:

CDP requires a free TCP port, so if you have port `9222` occupied, you may want just to start Chrome on another port:
```
testcafe 'chrome:port=9222'
```
vs
```
testcafe 'chrome:remote=localhost:9222'
```

**Comment by inikulin** [comment]
@AndreyBelym Still looks a bit messy for me. Let's ask others: \cc @DevExpress/testcafe

**Comment by miherlosev** [comment]
It is difficult to read and understand for me.
We have two feature requests: device emulation and headless mode.
We are implementing their using Chrome running via Chrome debug protocol.

So, browser provider name should tell about this.
```js
chrome:cdp;mobileEmulation={name=iPhone 4;orientation=horizontal}
chrome:cdp;headless
chrome:cdp;args={port:2222}
```
`cdp` suffix indicates that chrome run via CDP instead of normal run via command line
mobileEmulation and headless mode are most interesting options and they should be placed in top config level. 
`arg` - for all parameters as `host`, `port` and etc.

I use `mobileEmulation` term as in [ChromeDriver documentation](https://sites.google.com/a/chromium.org/chromedriver/mobile-emulation)

>I would like that CDB will be enabled by default, because it gives better screenshots: even if Chrome is in background, and without browser UI, on all platfroms (currently we don't include Chrome UI in screenshots only on Windows).

I don't think so. It is a big change. It's need to measure but I propose that Chrome running via CDP is slower than Chrome running via command line.

> testcafe "chrome:args=--disable-infobars --disable-session-crashed-bubble --kiosk;device.name=iPhone 

Some of these options we can add by default. For example, `--disable-infobars`.
Same as in https://github.com/DevExpress/testcafe-browser-tools/commit/9966ffa7f97498888623e65bd5c57d05ca4f6fa3

Aso I don't understand why we need `incognito` mode option.

**Comment by AlexanderMoskovkin** [comment]
I agree with @miherlosev in most points. Let's discuss it together today

**Comment by AndreyBelym** [comment]
[Summary updated](https://github.com/DevExpress/testcafe/pull/1461#issuecomment-299214557)

**Comment by AlexanderMoskovkin** [comment]
After one more review we want to propose more plain syntax for this features:

The common pattern is the following:
`chrome[:pathToBrowser][:headless][:emulation][:arg1=value;arg2=value]`

`path` - for the case if you want to run headless or device emulation mode for a portable browser;
`headless` - enable headless mode;
`emulation` - enable device emulation;

Available arguments:
- for device emulation: `width=<number>`, `height=<number>`, `density=<number>`, `mobile=<bool>`, `orientation=<vertical|horizontal>`, `userAgent=<string>`, `touch=<bool>`.
- cdp args: `cdpPort=<int>`

Here are some examples:
```sh
// Default headless mode
"chrome:headless"

// Device emulation
"chrome:emulation:device=iphone 6"
"chrome:emulation:width=100;height=200;density:1;mobile=false;orientation=vertical;userAgent=abc;touch=true"

// Headless device emulation
"chrome:headless:emulation:device=iphone 6;cdpPort=9223"

// Headless with cdpPort setting
"chrome:headless:cdpPort=9223"

// Headless/device emulation for a portable version
"chrome:path/to/chrome:headless:cdpPort=9223"
```

What do you think? /cc @DevExpress/testcafe

**Comment by miherlosev** [comment]
Fine

**Comment by AndreyBelym** [comment]
FPR

File: src/browser/provider/built-in/chrome.js [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+function createTempUserDataDir () {
+    tmp.setGracefulCleanup();
+
+    return tmp.dirSync({ unsafeCleanup: true });
+}
+
+export default {
+    openedBrowsers: {},
```
Comment (AlexanderMoskovkin) [inline src/browser/provider/built-in/chrome.js]: the file is huge. My proposal:
`built-in/chrome/index.js` - with provider definition
and several logical units in `built-in/chrome` with other functions

File: src/browser/provider/built-in/chrome.js [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        bounds.height += height - currentHeight;
+
+        await client.Browser.setWindowBounds({ windowId, bounds });
+    },
+
+    async hasCustomActionForBrowser (browserId) {
+        var { config, windowId, client } = this.openedBrowsers[browserId];
+
+        return {
+            hasResizeWindow:                !!client && (config.emulation || windowId || config.headless),
```
Comment (AlexanderMoskovkin) [inline src/browser/provider/built-in/chrome.js]: is it possible to call `resizeWindow` if run test in emulation mode? For example if we run in the iphone emulator how can we resize the window?

File: src/browser/provider/built-in/chrome/config.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+async function getNewConfig (configString) {
+    var { userArgs, modesString } = splitConfig(configString);
+    var { modes, optionsString }  = splitModes(modesString);
+    var options                   = splitOptions(optionsString, modes);
+
+    return Object.assign({ userArgs }, modes, options);
+}
+
+export default async function (configString) {
+    if (!configCache[configString])
```
Comment (AlexanderMoskovkin) [inline src/browser/provider/built-in/chrome/config.js]: let's add unit tests for this

File: src/browser/provider/built-in/chrome/local-chrome.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    var chromeOpenParameters = Object.assign({}, chromeInfo);
+
+    chromeOpenParameters.cmd = buildChromeArgs(config, cdpPort, chromeOpenParameters.cmd, tempUserDataDir);
+
+    await browserTools.open(chromeOpenParameters, pageUrl);
+}
+
+export async function stop ({ cdpPort }) {
+    if (!await killChrome(cdpPort))
```
Comment (AlexanderMoskovkin) [inline src/browser/provider/built-in/chrome/local-chrome.js]: add the comment why we kill twice

File: src/browser/provider/built-in/chrome/cdp.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        var { windowId } = await client.Browser.getWindowForTarget({ targetId: tab.id });
+
+        return windowId;
+    }
+    catch (e) {
+        return null;
+    }
+}
+
+async function setEmulationBounds (client, device, overrideWidth, overrideHeight) {
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/cdp.js]: `overrideWidth` and `overrideHeight` is a new bounds.
So, I propose to pack their to the single argument.
`async function setEmulationBounds (client, device, {newWidth, newHeight})`

File: src/browser/provider/built-in/chrome/index.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        Object.assign(runtimeInfo, cdpClientInfo);
+
+        this.openedBrowsers[browserId] = runtimeInfo;
+    },
+
+    async closeBrowser (browserId) {
+        var runtimeInfo = this.openedBrowsers[browserId];
+
+        if (cdp.canCloseBrowser(runtimeInfo))
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/index.js]: `canCloseBrowser` is unclear name.
Something like `isBrowserInEmulationMode` or `isInEmulation`.

File: src/browser/provider/built-in/chrome/config.js [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+const CONFIG_TERMINATOR_RE = /(\s+|^)-/;
+
+const HEADLESS_DEFAULT_WIDTH  = 1280;
+const HEADLESS_DEFAULT_HEIGHT = 800;
+
+var configCache = {};
+
+function hasMatch (array, re) {
+    return array.some(el => el.match(re));
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: Can we use `Array.find` for simplify queries?

File: src/browser/provider/built-in/chrome/config.js [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+        if (str[i] === '\\' && (str[i + 1] === '\\' || str [i + 1] === splitterChar))
+            i++;
+
+        result[result.length - 1] += str[i];
+    }
+
+    return result;
+}
+
+function splitConfig (str) {
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: `parseConfig`

File: src/browser/provider/built-in/chrome/config.js [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+    if (!configTerminatorMatch)
+        return { modesString: str, userArgs: '' };
+
+    return {
+        modesString: str.substr(0, configTerminatorMatch.index),
+        userArgs:    str.substr(configTerminatorMatch.index + configTerminatorMatch[1].length)
+    };
+}
+
+function splitModes (str) {
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: `parseMode`

File: src/browser/provider/built-in/chrome/config.js [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        orientation: orientation,
+        touch:       deviceData.capabilities.indexOf('touch') >= 0,
+        width:       deviceData.screen[orientation].width,
+        height:      deviceData.screen[orientation].height,
+        density:     deviceData.screen['device-pixel-ratio'],
+        userAgent:   deviceData['user-agent'],
+    };
+}
+
+function splitOptions (str, modes) {
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: `parseOptions`

File: src/browser/provider/built-in/chrome/config.js [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        touch:       touch,
+        mobile:      hasMatch(parsed, /^mobile$/) || isMatchTrue(parsed, /^mobile=(.*)/),
+        width:       Number(findMatch(parsed, /^width=(.*)/).match || NaN),
+        height:      Number(findMatch(parsed, /^height=(.*)/).match || NaN),
+        density:     Number(findMatch(parsed, /^density=(.*)/).match || NaN),
+        userAgent:   findMatch(parsed, /^userAgent=(.*)/).match,
+    };
+
+    Object.keys(specifiedDeviceOptions).forEach(key => {
+        if (!specifiedDeviceOptions[key] && specifiedDeviceOptions[key] !== 0 && specifiedDeviceOptions[key] !== false)
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: Remove empty string and undefined values

File: src/browser/provider/built-in/chrome/cdp.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    await remoteChrome.closeTab({ id: tab.id, port: cdpPort });
+}
+
+export async function takeScreenshot (path, { client, config }) {
+    var screenshot = await client.Page.captureScreenshot({ fromSurface: config.headless });
+
+    await fsWriteFile(path, screenshot.data, { encoding: 'base64' });
+}
+
+export async function resizeWindow (width, height, currentWidth, currentHeight, { config, client, windowId }) {
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/cdp.js]: `async function resizeWindow ({newWidth, newHeight}, {currentWidth, currentHeight}, { config, client, windowId })`

File: src/browser/provider/built-in/chrome/cdp.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
+import remoteChrome from 'chrome-remote-interface';
+import fs from 'fs';
+import Promise from 'pinkie';
+import pify from 'pify';
+
+
+const fsWriteFile = pify(fs.writeFile, Promise);
+
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/cdp.js]: Use `promisify` here from `src/utils/promisify.js` and below

Also we can move all promisified standard function to the separate file.
What do you think?

File: src/browser/provider/built-in/chrome/utils.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import fs from 'fs';
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/utils.js]: There are a several promisified functions - https://github.com/DevExpress/testcafe/search?utf8=%E2%9C%93&q=promisify%28&type=
Let's combine it to the separate file - `src/utils/promisified-functions.js`

File: src/browser/provider/built-in/chrome/config.js [hunk 15]
```diff
@@ -58,19 +56,21 @@ function splitConfig (str) {
...
 }
 
-function splitModes (str) {
-    var parsed   = splitEscaped(str, ':');
-    var pathMode = findMatch(parsed, /^path=(.*)/);
+function parseModes (str) {
+    var parsed         = splitEscaped(str, ':');
+    var pathModeRegExp = /^path=(.*)/;
+    var pathModeIndex  = findIndex(parsed, el => el.match(pathModeRegExp));
+    var pathModePath   = findMatch(parsed, pathModeRegExp);
```
Comment (miherlosev) [inline src/browser/provider/built-in/chrome/config.js]: Move path calcualtion to the separate function

File: src/browser/provider/built-in/chrome/cdp.js [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+async function setEmulationBounds (client, device, newDimensions) {
+    var width  = newDimensions ? newDimensions.width : device.width;
+    var height = newDimensions ? newDimensions.height : device.height;
+
+    await client.Emulation.setDeviceMetricsOverride({
+        width:             width,
+        height:            height,
+        deviceScaleFactor: device.density,
```
Comment (AlexanderMoskovkin) [inline src/browser/provider/built-in/chrome/cdp.js]: let's call it `scaleFactor`

**Comment by AndreyBelym** [comment]
@testcafe-build-bot \retest

**Comment by AndreyBelym** [comment]
FPR

**Pull request merged**